### ISSUE TITLE
Can I use synced-cron to time a single server event?

### ISSUE THREAD

**Issue by dalgard** [percolatestudio/meteor-synced-cron#41]
I'm building an auction system where the bidding on each item ends at a specific time. When an auction ends, certain operations should be executed. This is a one-time event, of course.

SyncedCron seems like the most robust solution for timed server events – is there an easy way of creating a job that runs once at a specific time in the future?

Thanks!

**Comment by dalgard** [comment]
I've reearched a bit deeper.

It seems that `synced-cron` expects there to always be a `next` event on the Later.js object – however, in some cases, Later.js doesn't have any more events. It may be that a schedule is only be active between two points in time, or – as in my case – simply *is a* point in time:

```javascript
SyncedCron.add({
  name: cron_name,
  schedule: function (parser) {
    // ending_at is a Date object set to some future date
    return parser.recur().on(ending_at).fullDate();
  },
  job: function () {
    console.log("hep!");
    // remove job so it doesn't recur
    SyncedCron.remove(cron_name);
  }
});
```

The above code results in the following exception:

```
Exception in setTimeout callback: TypeError: Cannot call method 'getTime' of undefined
     at scheduleTimeout (packages/percolate:synced-cron/synced-cron-server.js:240:1)
```

... since in the `synced-cron` code, the call to `getTime` fails when there are no more events. Is there a specific point in my code I could call `SyncedCron.remove` before the failing call is made?

https://github.com/percolatestudio/meteor-synced-cron/blob/125ae97a05fbc27831762e239ecce6cf5fd97555/synced-cron-server.js#L240

Other ways to make this work? Would be easy to patch `synced-cron` to stop when there are no `next` events, but I think this is an API decision.

**Comment by dalgard** [comment]
So, I created a patch... What it does is simply removing the job if the schedule doesn't have any more events. Also, it makes sure the next event isn't triggered if the entry is removed inside the job function body.

**Comment by tylerstraub** [comment]
I ran into a similar problem on one of my projects, I opted to go with the CRON parser (parser.cron) which accepts standard CRON syntax. From there I was able to formulate a single event using an expression such as; "10 18 26 4 ? 2015"

**Comment by zol** [comment]
Thanks for the input @tylerstraub . My plan is still to take a look at this ticket as soon as I have some free time.

**Comment by dalgard** [comment]
Don't know how much the code has developed, but my pull request solving this problem is quite simple – you could just use that :) I'm using it in production, anyways.

**Comment by zol** [comment]
@dalgard - The reason we ask that PR authors write tests is because we maintain quite a few o/s projects and don't necessarily have the bandwidth to go through each PR in detail and support it into the future. It's actually not that we're in doubt whether the code is production ready (we give you guys the benefit of doubt). If the PR comes with tests however, we're comfortable merging it with less scrutiny knowing that the author has described exactly what the functionality does and in fact we won't break it in future versions.

This way we can support functionality that we don't actually use ourselves as the tests become the vehicle that ensures we don't break stuff that others rely on.

In short, if we don't use something ourselves and there are no tests for it there is a pretty good chance that functionality may break in between releases - this is something we obviously want to avoid.

Hence with PR's like yours that don't come with tests, we need to schedule in the time and come up to speed with the functionality + write tests ourselves. Unfortunately, during busy times this falls down our priority queue.

**Comment by dalgard** [comment]
I totally understand and earnestly consider it a healthy priority – had it been my project, I hope I would be equally methodical. All I would do, is encourage you to review my PR and see how limited the adjustments would be for resolving this issue. Throwing together a patch on the basis of the lines I've submitted is a convenient path at times where professional obligations and other things take rightful precedence :)

**Comment by BradRyan** [comment]
Any update on this.  Just noticed this issue on my project.  It actually causes my server to crash and restart.  In the meantime I'll see about adopting the solution that @dalgard came up with.

**Comment by BradRyan** [comment]
FYI there's an orphaned package which uses dalgard's solution.  I've been using it... so far so good.
https://atmospherejs.com/orphan/synced-cron

**Comment by sylido** [comment]
+1 for merge.

**Comment by rb365** [comment]
+1 for merge

**Comment by elie222** [comment]
+1 for tests to be added + merge

**Comment by davecromberge** [comment]
+1 for merge.

**Comment by KyleKing** [comment]
+1 as well!

**Comment by zol** [comment]
Released a fix to point in time schedules in version 1.3.0. Let me know if there are any issues with it. Thanks for your patience.

**Comment by yyl** [comment]
I am using version 1.3.2, and want to fire a once-off job, but still facing the same issue:

        SyncedCron.add({
          name: 'reminder_email_' + uid,
          schedule: function(parser) {
            return parser.recur().on(new Date() + 60000).fullDate();
          },
          job: function() {
            // functions here
          }
        });

Console prints:

```
I20160519-17:56:27.539(-4)? Exception while invoking method 'addTask' TypeError: Cannot call method 'getTime' of undefined
I20160519-17:56:27.539(-4)?     at scheduleTimeout (packages/percolate_synced-cron/synced-cron-server.js:309:1)
I20160519-17:56:27.540(-4)?     at Object.SyncedCron._laterSetTimeout (packages/percolate_synced-cron/synced-cron-server.js:289:1)
I20160519-17:56:27.540(-4)?     at Object.SyncedCron._laterSetInterval (packages/percolate_synced-cron/synced-cron-server.js:257:1)
I20160519-17:56:27.540(-4)?     at scheduleEntry (packages/percolate_synced-cron/synced-cron-server.js:104:1)
I20160519-17:56:27.540(-4)?     at Object.SyncedCron.add (packages/percolate_synced-cron/synced-cron-server.js:127:1)
I20160519-17:56:27.541(-4)?     at [object Object].addTask (server/methods.js:107:20)
I20160519-17:56:27.541(-4)?     at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1704:12)
I20160519-17:56:27.541(-4)?     at packages/ddp-server/livedata_server.js:711:19
I20160519-17:56:27.541(-4)?     at [object Object]._.extend.withValue (packages/meteor/dynamics_nodejs.js:56:1)
I20160519-17:56:27.541(-4)?     at packages/ddp-server/livedata_server.js:709:40
```

**Comment by adrianbadarau** [comment]
@yyl The cron is working as it should, the on method just requires a date not a timestamp.
The correct way to initialise the cron, if you just wanted to add 6s to the current date-time then you could just do something like:

`    SyncedCron.add({
        name:"Test single time cron",
        schedule(parser){
            let time = new Date(Date.now() + 6000);
            return parser.recur().on(time).fullDate();
        },
        job(){
            console.log("IT IS WORKING")
        }
    });`

**Comment by yyl** [comment]
@adrianbadarau great that works, thanks!
